imageops: introduce tile function

MicroJoe opened the pull request:
I license past and future contributions under the dual MIT/Apache-2.0 license,
allowing licensees to chose either at their option.

Example usage:

```rust
pub fn example_tile() {
    let mut img = RgbaImage::new(300, 170);
    let background_color = Rgba::from_slice(&[20, 20, 20, 255]);
    for p in img.pixels_mut() {
        *p = *background_color;
    }

    let tile = image::open("spinner.png").unwrap();

    image::imageops::tile(&mut img, &tile);
    img.save("test.png").unwrap();
}
```

Example output image:

![test](https://user-images.githubusercontent.com/989521/87873676-c8a0ab80-c9c3-11ea-9d0d-e15105da5a4a.png)

HeroicKatora commented on src/imageops/mod.rs:
```diff
@@ -164,6 +164,21 @@ where
     }
 }
 
+/// Tile an image by repeating it multiple times
```
Could you add the example here? Marked with `no_run` since it tries to load from disk.

HeroicKatora commented on src/imageops/mod.rs:
```diff
@@ -164,6 +164,21 @@ where
+    I: GenericImage,
+    J: GenericImageView<Pixel = I::Pixel>,
+{
+    for i in 0..(bottom.width() / top.width() + 1) {
+        let x = i * top.width();
```
If both images have the same width then this loop runs twice, the second time not adding any pixels. You can use the iterator combinators to avoid this off-by-one:

```suggestion
    for x in (0..bottom.width()).step_by(top.width()) {
```

HeroicKatora commented:
@MicroJoe Do you you'll get to addressing the remarks this weekend? No pressure, I'm just trying to plan the potential `0.23.8` release.

MicroJoe commented:
> @MicroJoe Do you think you'll get to addressing the remarks this weekend? No pressure, I'm just trying to plan the potential `0.23.8` release.

Thanks for the reminder, applied proposed changes as well as added an example. Please let me know if anything else is missing.

By the way, my setup did want to reformat the whole `mod.rs` using `rustfmt` as it looks like this file was making rustfmt unhappy. Do you know if there is any plan to force apply rustfmt on the whole project?

HeroicKatora commented:
Thank you!

HeroicKatora commented:
> By the way, my setup did want to reformat the whole mod.rs using rustfmt as it looks like this file was making rustfmt unhappy. Do you know if there is any plan to force apply rustfmt on the whole project?

Certainly not 'the default' settings. It is quite opinionated and that in many regards can decrease legibility of code. [See here for some more](https://github.com/image-rs/image/pull/1162#pullrequestreview-374794859). That said, it does come up often enough that new contributors have auto-formatting on save enabled and then it can be harder to contribute. If there is some way to _selectively_ enable only small parts of the formatting (and disable the rest) through a configuration file, I'd certainly be interested.
